syft/messaging/plan.py

```diff
@@ -104,111 +95,256 @@ def append(self, key):
-                for param in ptr.parameters():
-                    id_at_location.append(param.id_at_location)
-            else:
-                id_at_location.append(ptr.id_at_location)
-
-        return id_at_location
+        tensors = []
+        for state_id in self.state_ids:
+            tensor = self.owner.get_obj(state_id)
+            assert tensor.id == state_id
+            tensors.append(tensor)
+        return tensors
+
+    def clone_state(self) -> Dict:
+        """
+        Return a clone of the state elements. Tensor ids are kept.
+        """
+        return {tensor.id: tensor.clone() for tensor in self.tensors()}
 
-    def copy(self):
-        copied_state = {}
-        for key in self.keys:
-            t = getattr(self.plan, key)
-            copied_state[key] = t.copy()
-        return copied_state
+    def copy(self) -> "State":
+        state = State(owner=self.owner, keys=self.keys.copy(), state_ids=self.state_ids.copy())
+        return state
 
     def read(self, key):
         """
-        Returns one element referenced by the store
+        Returns one element referenced by the store of keys.
+        Note that this only works for plans built locally
         """
         return getattr(self.plan, key)
 
-    def set_(self, dict_state):
+    def set_(self, state_dict):
         """
         Reset inplace the state by feeding it a dict of tensors or params
         """
-        for key in self.keys:
-            delattr(self.plan, key)
+        assert list(self.state_ids) == list(state_dict.keys())
 
-        for key in dict_state.keys():
-            setattr(self.plan, key, dict_state[key])
+        for state_id, new_tensor in state_dict.items():
+            tensor = self.owner.get_obj(state_id)
 
-        self.keys = list(dict_state.keys())
+            with torch.no_grad():
+                tensor.set_(new_tensor)
 
-    def is_missing_grad(self, t):
-        return isinstance(t, torch.nn.Parameter) and t.grad is None
+            tensor.child = new_tensor.child if new_tensor.is_wrapper else None
+            tensor.is_wrapper = new_tensor.is_wrapper
+            if tensor.child is None:
+                delattr(tensor, "child")
 
-    def create_grad_objects(self, p):
-        o = p.sum()
-        o.backward()
-        if p.grad is not None:
-            p.grad -= p.grad
+    @staticmethod
+    def create_grad_if_missing(tensor):
+        if isinstance(tensor, torch.nn.Parameter) and tensor.grad is None:
+            o = tensor.sum()
+            o.backward()
+            if tensor.grad is not None:
+                tensor.grad -= tensor.grad
+
+    def send_for_build(self, location, **kwargs):
+        """
+        Send functionality that can only be used when sending the state for
+        building the plan. Other than this, you shouldn't need to send the
+        state separately.
+        """
+        assert location == self.plan  # ensure this is a send for the build
```

Comment:
I didn't follow everything but it looks a bit weird to compare a plan and a location

Reply:
Ok I'll change it to id comparison
Actually plans and workers are kinds of cousins: when you build a plan you send tensors and commands to the plans, and he stores the commands (=they form the "readable_plan") but doesn't run them.
Look some methods like _recv_msg are held by both

Reply:
Oh yes, I had missed that part where we removed owner from the plan, etc.!
How will that work when we use distributed workers for real? Where will plans be?

Reply:
Same thing, nothing changes: at building (which always happens locally) you alway send evrything to the plan just like if it was a Virtual worker, then you can run a simulated computation, and that plan/worker stores all commands, which gonna make the readable_plan.
This readable_plan is included in the state attributes and is then sent to real remote workers for remote execution